spring-integration-ftp/src/main/java/org/springframework/integration/ftp/session/FtpSession.java

```diff
@@ -200,18 +200,35 @@ public boolean rmdir(String directory) throws IOException {
 	public boolean exists(String path) throws IOException {
 		Assert.hasText(path, "'path' must not be empty");
 
-		String currentWorkingPath = this.client.printWorkingDirectory();
-		Assert.state(currentWorkingPath != null,
-				"working directory cannot be determined, therefore exists check can not be completed");
 		boolean exists = false;
-
+		InputStream inputStream = null;
 		try {
-			if (this.client.changeWorkingDirectory(path)) {
-				exists = true;
-			}
+			inputStream = this.client.retrieveFileStream(path);
```

Comment:
H-m. The problem with new `fileExists()` and `dirExists()` that they are needed only for FTP, meanwhile `LSTAT` is enough for SFTP and `getObjectMetadata()` for AWS S3.

So, may be that isn't so big deal to `changeWorkingDirectory()` twice for FTP server as a fallback option...